cmd/setup_darwin.go

```diff
@@ -17,13 +17,158 @@
-	fmt.Println("")
+// isRoot returns true if weep is running as root and false otherwise
+func isRoot() bool {
+	return os.Geteuid() == 0
+}
+
+// writeFileFromEmbedded calls the appropriate file writing function based on the value of commit
+func writeFileFromEmbedded(prefix string, filename string, commit bool) error {
+	data, err := SetupExtras.ReadFile(prefix + filename)
+	port := viper.GetString("server.port")
+	data = bytes.Replace(data, []byte("WEEP_PORT"), []byte(port), -1)
+	if err != nil {
+		return err
+	}
+	if commit {
+		err = writeFileGo(filename, data)
+		return err
+	} else {
+		err = writeFileShell(filename, data)
+		return err
+	}
+}
+
+// writeFileGo creates a backup of the target file and writes new content using Go
+func writeFileGo(filename string, data []byte) error {
+	print("writing ", filename, "...\n")
+	// Ignore error on backup, we're just trying to be nice anyway
+	_ = backupFile(filename)
+	err := ioutil.WriteFile(filename, data, 0644)
+	return err
+}
+
+// backupFile creates a copy of filename with a timestamp appended to the filename
+func backupFile(filename string) error {
+	src := filename
+	dst := fmt.Sprintf("%s.%s", filename, time.Now().Format("20060102150405"))
+	sourceFileStat, err := os.Stat(src)
+	if err != nil {
+		return err
+	}
+
+	if !sourceFileStat.Mode().IsRegular() {
+		return fmt.Errorf("%s is not a regular file", src)
+	}
+
+	source, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer source.Close()
+
+	destination, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer destination.Close()
+	_, err = io.Copy(destination, source)
+	return err
+}
+
+// writeFileGo prints a script to create a backup of the target file and writes new content
+func writeFileShell(filename string, data []byte) error {
+	// make a backup copy, ignoring failure (e.g. if source file doesn't exist)
+	fmt.Printf("cp %s %s.$(date +%%Y%%m%%d%%H%%M%%S) || true\n", filename, filename)
+	fmt.Printf("cat << EOF > %s\n", filename)
+	fmt.Print(string(data))
+	fmt.Print("EOF\n")
+	return nil
+}
+
+// reloadPlist uses executes launchctl to unload and load the specified plist, or prints the commands
+// to do so if commit is false
+func reloadPlist(plistFile string, commit bool) error {
+	unloadCmd := exec.Command("launchctl", "unload", plistFile)
+	loadCmd := exec.Command("launchctl", "load", plistFile)
+	if commit {
+		print("loading ", plistFile, "...\n")
+		// Ignore error on unload because this plist might not exist
+		_ = unloadCmd.Run()
+		err := loadCmd.Run()
+		return err
+	}
+	fmt.Println(unloadCmd.String(), " || true")
+	fmt.Println(loadCmd.String())
+	return nil
+}
+
+func Setup(cmd *cobra.Command, commit bool) error {
+	if commit && !isRoot() {
+		cmd.Print("not running as root. If this fails, try again using sudo.\n")
+	}
+	if !commit {
+		fmt.Println("#!/bin/sh")
```

Comment:
Do you think building the string and printing at the very end might be a better idea, in the case an error occurs?
Since doing it this way, we end up with things like:
```
#!/bin/sh
Valid instructions

Error message
exit 1
```
which may be slightly confusing and users in a rush, or using pipe may end up executing commands that are not valid

Reply:
Yeah, that's a good point. It'll take a bit of refactoring but that seems to be the right approach.